### PR TITLE
[visq] Introduce 'batch' mode

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit introduces 'batch' mode for `visq` component.

Draft: #10432
Related to: #10393
ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: compiler/visq/visq [hunk 1]
```diff
@@ -107,6 +118,64 @@ def _run_dalgona(model, data, analysis, save_dir):
         raise
 
 
+def save_inputs(inputs, inputs_path):
```
Comment (jinevening) [inline compiler/visq/visq:121]:
```suggestion
# Generate h5 file that contains a dataset of a single batch
# This is for batch execution of visq
def gen_batch_h5(inputs_data, inputs_path):
```

File: compiler/visq/visq [hunk 2]
```diff
@@ -107,6 +118,64 @@ def _run_dalgona(model, data, analysis, save_dir):
...
+    for i in range(len(inputs)):
+        sample = group.create_group(str(i))
+        for j in range(len(inputs[i])):
+            sample.create_dataset(str(j), data=inputs[i][j])
+
+    h5_file.close()
+    return output_path
+
+
+def advance_on_data(models, data_batch_path, computers):
```
Comment (jinevening) [inline compiler/visq/visq:137]: Can you split models into fp32_model and fq_model?
```suggestion
def advance_on_data(models, data_batch_path, computers):
```

File: compiler/visq/visq [hunk 3]
```diff
@@ -107,6 +118,64 @@ def _run_dalgona(model, data, analysis, save_dir):
...
+            computers[metric_key][0].advance_on(fp32_dir, fq_dir)
+
+
+def _run_batch(models, data, computers, batch_size):
+    with tempfile.TemporaryDirectory() as inputs_dir:
+        with h5.File(data, 'r') as f:
+            dataset = f['value']
+
+            inputs = []
+            for index, data_index in enumerate(dataset):
```
Comment (jinevening) [inline compiler/visq/visq:161]: index is not used anywhere. You don't have to use enumerate.

File: compiler/visq/visq [hunk 4]
```diff
@@ -107,6 +118,64 @@ def _run_dalgona(model, data, analysis, save_dir):
...
+                inputs.append(cur_inputs)
+                if len(inputs) >= batch_size:
+                    input_path = save_inputs(inputs, inputs_dir)
+                    advance_on_data(models, input_path, computers)
+                    inputs = []
+
+            if len(inputs) > 0:
+                input_path = save_inputs(inputs, inputs_dir)
+                advance_on_data(models, input_path, computers)
+                inputs = []
```
Comment (jinevening) [inline compiler/visq/visq:176]: Seems unnecessary
```suggestion
```

**Comment by jinevening** [comment]
Please fix the format error.

**Comment by stamalakhov** [comment]
> Please fix the format error.

Sorry for delay. Was out of office.

**Review by jinevening — DISMISSED** [review]
LGTM

File: compiler/visq/visq [hunk 5]
```diff
@@ -107,6 +117,66 @@ def _run_dalgona(model, data, analysis, save_dir):
...
+def advance_on_data(fp32_model, fq_model, data, computers):
+
+    curr_dir = Path(__file__).parent.resolve()
+    dump_fp32_py = curr_dir / 'visqlib' / 'DumpFP32FM.py'
+    dump_fq_py = curr_dir / 'visqlib' / 'DumpFakeQuantFM.py'
+
+    with tempfile.TemporaryDirectory() as fp32_dir, \
+         tempfile.TemporaryDirectory() as fq_dir:
+
+        #print("advance on data + {fp_dir} + {fq_dir}".format(fp_dir = fp32_dir, fq_dir = fq_dir))
```
Comment (seanshpark) [inline compiler/visq/visq:148]: Q) what is the need for this comment?

File: compiler/visq/visq [hunk 6]
```diff
@@ -172,19 +242,48 @@ def _save_dot(circle_path: str, dot_path: str, metric: str, colors: list, qerror
...
+                          batch_size):
 
-    fp32_model = args.fp32_circle
-    q_model = args.q_circle
-    data = None
-    if valid_attr(args, 'data'):
-        data = args.data
-    dump_dot_graph = args.dump_dot_graph
+    with tempfile.TemporaryDirectory() as model_dir:
+        fq_model = model_dir + '/fq_model.circle'
```
Comment (seanshpark) [inline compiler/visq/visq:249]: is there a reason to fix this name?

File: compiler/visq/visq (hunk 5 above)
Comment (stamalakhov) [reply compiler/visq/visq:148]: Sorry. No. We don't need it. I'll remove it.

File: compiler/visq/visq (hunk 6 above)
Comment (stamalakhov) [reply compiler/visq/visq:249]: It stands for `fake_quantize_model`. As it's an internal it should be created in temporary directory to be removed as soon as possible.  IMHO there is no other reason to call it `fq_model` other than ease of use and for the full batch it's also called the same.

Comment (seanshpark) [reply compiler/visq/visq:248]: OK, it is temporary file.

Comment (seanshpark) [reply compiler/visq/visq:248]: maybe use `TemporaryFile()` not a fixed name?

Comment (stamalakhov) [reply compiler/visq/visq:248]: Yes, IMHO `TemporaryFile()` can also be used, but then IMHO we should also change it for `run_on_data`. IMHO this is not needful for this PR. May be could deal with later?

Comment (seanshpark) [reply compiler/visq/visq:248]: OK got it

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]

**Pull request merged**